cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/sync/class-storage.php

```diff
@@ -0,0 +1,242 @@
+<?php
+/**
+ * Storage management options.
+ *
+ * @package Cloudinary
+ */
+
+namespace Cloudinary;
+
+namespace Cloudinary\Sync;
+
+use Cloudinary\Component\Notice;
+use Cloudinary\Sync;
+
+/**
+ * Class Filter.
+ *
+ * Handles filtering of HTML content.
+ */
+class Storage implements Notice {
+
+	/**
+	 * Holds the Plugin instance.
+	 *
+	 * @since   0.1
+	 *
+	 * @var     \Cloudinary\Plugin Instance of the plugin.
+	 */
+	protected $plugin;
+
+
+	/**
+	 * Holds the Plugin Media instance.
+	 *
+	 * @since   0.1
+	 *
+	 * @var     \Cloudinary\Media Instance of the media object.
+	 */
+	protected $media;
+
+	/**
+	 * Holds the Sync instance.
+	 *
+	 * @since   0.1
+	 *
+	 * @var     \Cloudinary\Sync Instance of the plugin.
+	 */
+	protected $sync;
+
+	/**
+	 * Holds the Download Sync instance.
+	 *
+	 * @since   0.1
+	 *
+	 * @var     \Cloudinary\Sync\Download_Sync Instance of the plugin.
+	 */
+	protected $download;
+
+	/**
+	 * Holds the Connect instance.
+	 *
+	 * @since   0.1
+	 *
+	 * @var     \Cloudinary\Connect Instance of the plugin.
+	 */
+	protected $connect;
+
+	/**
+	 * Holds an array of the storage settings.
+	 *
+	 * @var array
+	 */
+	protected $settings;
+
+	/**
+	 * Filter constructor.
+	 *
+	 * @param \Cloudinary\Plugin $plugin The plugin.
+	 */
+	public function __construct( \Cloudinary\Plugin $plugin ) {
+		$this->plugin   = $plugin;
+		$this->sync     = $this->plugin->components['sync'];
+		$this->media    = $this->plugin->components['media'];
+		$this->download = $this->sync->managers['download'];
+
+		add_action( 'cloudinary_register_sync_types', array( $this, 'setup' ), 20 );
+	}
+
+	/**
+	 * Generate a signature for this sync type.s
+	 *
+	 * @return string
+	 */
+	public function generate() {
+		return $this->settings['offload'] . $this->settings['low_res'];
+	}
+
+	/**
+	 * Process the storage sync for an attachment.
+	 *
+	 * @param int $attachment_id The attachment ID.
+	 */
+	public function sync( $attachment_id ) {
+
+		switch ( $this->settings['offload'] ) {
+			case 'cld':
+				$this->remove_local_assets( $attachment_id );
+				break;
+			case 'dual_low':
+				$url = $this->media->cloudinary_url( $attachment_id, 'full', array( array( 'effect' => 'blur:100', 'quality' => $this->settings['low_res'] . ':440' ) ), null, false, true );
+				break;
+			case 'dual_full':
+				$state = $this->media->get_post_meta( $attachment_id, Sync::META_KEYS['storage'], true );
+				if ( ! empty( $state ) && 'dual_full' !== $state ) {
+					// Only do this is it's changing a state.
+					$url = $this->media->cloudinary_url( $attachment_id, '', array(), null, false, false );
+				}
+				break;
+		}
```

Comment:
We need to pick a `default` state or we open the system up for breaking (ie. admin with malicious intent)